test/Tester/Lifecycle/LifecycleTests.cs

```diff
@@ -0,0 +1,229 @@
+﻿
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans;
+using Xunit;
+
+namespace Tester
+{
+    public class LifecycleTests
+    {
+        [Fact, TestCategory("BVT"), TestCategory("Lifecycle")]
+        public async Task FullLifecycleTest()
+        {
+            const int observersPerStage = 2;
+            Dictionary<TestStages, int> observerCountByStage = new Dictionary<TestStages, int>();
+            foreach (TestStages stage in Enum.GetValues(typeof(TestStages)))
+            {
+                observerCountByStage[stage] = observersPerStage;
+            }
+            Dictionary<TestStages, List<Observer>> observersByStage = await RunLifeCycle(observerCountByStage, null, null);
+
+            Assert.Equal(observerCountByStage.Count, observersByStage.Count);
+            foreach (KeyValuePair<TestStages,List<Observer>> kvp in observersByStage)
+            {
+                Assert.Equal(observerCountByStage[kvp.Key], kvp.Value.Count);
+                Assert.True(kvp.Value.All(o => o.Started));
+                Assert.True(kvp.Value.All(o => o.Stopped));
+                Assert.True(kvp.Value.All(o => !o.FailedOnStart));
+                Assert.True(kvp.Value.All(o => !o.FailedOnStop));
+            }
+        }
+
+        [Fact, TestCategory("BVT"), TestCategory("Lifecycle")]
+        public async Task FailOnStartOnEachStageLifecycleTest()
+        {
+            const int observersPerStage = 2;
+            Dictionary<TestStages, int> observerCountByStage = new Dictionary<TestStages, int>();
+            foreach (TestStages stage in Enum.GetValues(typeof(TestStages)))
+            {
+                observerCountByStage[stage] = observersPerStage;
+            }
+
+            foreach (TestStages stage in Enum.GetValues(typeof(TestStages)))
+            {
+                Dictionary<TestStages, List<Observer>> observersByStage = await RunLifeCycle(observerCountByStage, stage, null);
+
+                Assert.Equal(observerCountByStage.Count, observersByStage.Count);
+                foreach (KeyValuePair<TestStages, List<Observer>> kvp in observersByStage)
+                {
+                    Assert.Equal(observerCountByStage[kvp.Key], kvp.Value.Count);
+                    if (kvp.Key < stage)
+                    {
+                        Assert.True(kvp.Value.All(o => o.Started));
+                        Assert.True(kvp.Value.All(o => o.Stopped));
+                        Assert.True(kvp.Value.All(o => !o.FailedOnStart));
+                        Assert.True(kvp.Value.All(o => !o.FailedOnStop));
+                    } else if (kvp.Key == stage)
```

Comment:
NIT since it's in the test: `else` statement should start on a new line. Same after